modules/renter/proto/contractset.go

```diff
@@ -0,0 +1,159 @@
+package proto
+
+import (
+	"sync"
+
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+// A safeContract protects a RenterContract with a mutex.
+type safeContract struct {
+	modules.RenterContract
+	mu sync.Mutex
+}
+
+// A ContractSet provides safe concurrent access to a set of contracts. Its
+// purpose is to serialize modifications to individual contracts, as well as
+// to provide operations on the set as a whole.
+type ContractSet struct {
+	contracts map[types.FileContractID]*safeContract
+	mu        sync.Mutex
+
+	// caches (initialized by initCache)
+	flatIDs       []types.FileContractID
+	flatContracts []modules.RenterContract
+	once          sync.Once
+}
+
+// initCache populates the flatIDs and flatContracts fields so that repeated
+// work can be avoided when calling the IDs and Contracts methods. It must be
+// called under lock. initCache should be called by IDs and Contracts if a
+// call to Insert, Return, or Delete has invalidated the cache. This is
+// managed via a sync.Once.
+func (cs *ContractSet) initCache() {
```

Comment:
I'm not sure that `init` is the clearest prefix for this func. Initialization usually refers to something that is done once, but this func is used to construct and reconstruct the cache in subsequent calls to IDs and Contracts.